code/asteroid/asteroid.cpp

```diff
@@ -733,52 +766,57 @@ static void asteroid_aim_at_target(object *objp, object *asteroid_objp, float de
 }
 
 /**
- * Call once per frame to maybe throw an asteroid at a ship.
+ * Call once per frame to maybe throw some asteroids at one or more ships.
  *
- * @param count asteroids already targeted on
  */
-static void maybe_throw_asteroid(int count)
+static void maybe_throw_asteroid()
 {
-	if (!timestamp_elapsed(Next_asteroid_throw)) {
-		return;
-	}
+	//sanitize the list
+	for (size_t i = 0; i < Asteroid_targets.size();) {
+		object* target_objp = &Objects[Asteroid_targets[i].objnum];
+		if (target_objp->type == OBJ_NONE || target_objp->signature != Asteroid_targets[i].signature) {
+			Asteroid_targets[i] = Asteroid_targets.back();
+			Asteroid_targets.pop_back();
+		} else  // if we needed to cull we should not advance because we just moved a new asteroid target into this spot
+			i++;
+	}
+
+	for (asteroid_target& target : Asteroid_targets) {
+		if (!timestamp_elapsed(target.throw_stamp))
+			continue;
 
-	if (Asteroid_throw_objnum == -1) {
-		return;
-	}
+		object* target_objp = &Objects[target.objnum];
+		if (!asteroid_is_ship_inside_field(&Asteroid_field, &target_objp->pos, target_objp->radius))
+			continue;
+		
+		// This should've been greater equal, but alas...
+		if (target.incoming_asteroids > The_mission.ai_profile->max_incoming_asteroids[Game_skill_level])
+			continue;
 
-	nprintf(("AI", "Incoming asteroids: %i\n", count));
+		nprintf(("AI", "Incoming asteroids to %s: %i\n", Ships[target_objp->instance].ship_name, target.incoming_asteroids));
 
-	if (count > The_mission.ai_profile->max_incoming_asteroids[Game_skill_level])
-		return;
+		target.throw_stamp = timestamp(1000 + 1200 * target.incoming_asteroids /(Game_skill_level+1));
 
-	Next_asteroid_throw = timestamp(1000 + 1200 * count/(Game_skill_level+1));
+		int subtype = Random::next(NUM_DEBRIS_POFS);
 
-	ship_obj	*so;
-	for ( so = GET_FIRST(&Ship_obj_list); so != END_OF_LIST(&Ship_obj_list); so = GET_NEXT(so) ) {
-		object *A = &Objects[so->objnum];
-		if (so->objnum == Asteroid_throw_objnum) {
-			int subtype = Random::next(NUM_DEBRIS_POFS);
-			while (Asteroid_field.field_debris_type[subtype] == -1) {
-				subtype = (subtype + 1) % NUM_DEBRIS_POFS;
-			}
-			object *objp = asteroid_create(&Asteroid_field, ASTEROID_TYPE_LARGE, subtype);
-			if (objp != NULL) {
-				asteroid_aim_at_target(A, objp, ASTEROID_MIN_COLLIDE_TIME + frand() * 20.0f);
+		while (Asteroid_field.field_debris_type[subtype] == -1)
```

Comment:
Is there a sanity check somewhere to make sure there is at least one asteroid subtype that isn't -1?

Reply:
It appears FRED will not let you leave the dialog without at least one selected, but still, better safe than sorry...